Development practices/Turbo frames.md

```diff
@@ -0,0 +1,315 @@
+
+That's because the form was submitted from the frame with the `new_author` id.
+Now the AuthorsController#create action renders the Authors#index page that contains an empty frame with the same `new_author` id.
+
+So basically the only change on the page is that the form for creating a new author will disappear because it is replaced with the empty frame.
+The rest of the page, including the authors list will remain intact and that's why the newly created author won't be visible until the page is refreshed.
+
+There is a workaround for that, we can wrap the whole index page inside of another turbo frame:
+```ruby
+# index.html.slim
+
+= turbo_frame_tag 'main'
+  h2 = 'Authors'
+  = link_to 'Add author', new_author_path, data: { 'turbo-frame': dom_id(Author.new) }
+
+  = turbo_frame_tag Author.new
+  - @authors.each do |author|
+    = turbo_frame_tag "edit_author_#{author.id}"
+      .row
+        .col
+          = author.first_name
+        .col
+          = author.last_name
+        .col
+          = link_to 'Edit', edit_author_path(author.id)
+```
+
+Then the view for the #new action can be updated like this:
+```ruby
+# new.html.slim
+
+= turbo_frame_tag Author.new
+  = simple_form_for @author, data: { turbo_frame: 'main' } do |f|
+    .row
+      .col
+        = f.input :first_name
+      .col
+        = f.input :last_name
+      .col
+        = f.button :submit
+```
+
+This way, when we submit the form with the valid attributes, it will act as if it has been submitted from the `main` turbo frame because of the `data-turbo-frame` attribute.
+The AuthorsController#create action then renders the Authors#index page, and replaces the content from the `main` turbo frame, which almost acts like the whole page has been refreshed.
+Now the newly created author should be visible in the authors list.
+
+Another way of achieving this behaviour is by using [Turbo streams](https://turbo.hotwired.dev/handbook/streams).
+
+### target
+The documentation for the turbo frames describes the `target` attribute like this:
+
+*By default, navigation within a frame will target just that frame. This is true for both following links and submitting forms. But navigation can drive the entire page instead of the enclosing frame by setting the target to _top. Or it can drive another named frame by setting the target to the ID of that frame.*
+
+Setting the target to `_top` does actually drive the entire page:
+```Html
+<turbo-frame id='authors' target='_top'>
+  <a href='/authors/1'>
+    Following link will replace the whole page, not just this frame.
+  </a>
+</turbo-frame>
+```
+
+However, the second part (*Or it can drive another named frame by setting the target to the ID of that frame.*) doesn't seem to work in development without using turbo streams ([discussion](https://discuss.hotwired.dev/t/target-attribute-on-turbo-frame-tags/1959)) and `data-turbo-frame` attribute can be used to achieve the required functionality.
```

Comment:
There's an example on Atlas using `target:` that works, you just have to put the target on both turbo frames.
Your previous example could be rewritten as:
```ruby
# index.html.slim

= turbo_frame_tag 'main'
  h2 = 'Authors'
  = link_to 'Add author', new_author_path, data: { 'turbo-frame': dom_id(Author.new) }

  = turbo_frame_tag Author.new, target: 'main' # <--
  - @authors.each do |author|
    = turbo_frame_tag "edit_author_#{author.id}"
      .row
        .col
          = author.first_name
        .col
          = author.last_name
        .col
          = link_to 'Edit', edit_author_path(author.id)

# new.html.slim

= turbo_frame_tag Author.new, target: 'main' # <--
  = simple_form_for @author do |f|
    .row
      .col
        = f.input :first_name
      .col
        = f.input :last_name
      .col
        = f.button :submit
```